### PR TITLE
feat: Add App Config Feature Flag type

### LINKED ISSUE #23720: Expose App Config Configuration Profile Types
AWS App Config has released the Feature Flags service to GA, as of 2022/03/15. This is available already in the Go SDK, but isn't exposed in the Terraform provider for the App Config service.

The `Type` property could be exposed in the Configuration Profile function that is creating the config profile

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* aws_XXXXX

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

### PR DESCRIPTION
**Pull request by ryancormack**
Feature Flags in App Config are now GA, https://aws.amazon.com/about-aws/whats-new/2022/03/aws-appconfig-feature-flags/

This Pull Request exposes the ability to set the Configuration Profile type to either Free Form, or Feature Flag, as per the AWS API docs, https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_CreateConfigurationProfile.html#appconfig-CreateConfigurationProfile-request-Type

I've marked the default value to be FreeForm. Whilst this isn't explicitly documented on AWS, this is the observed behaviour from creating an App Config Profile without setting this value. Using this default maintains backwards compatibility. The validation function uses the regex function documented by AWS

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #23720

Output from acceptance testing:

```
$ make testacc TESTS=TestAccXXX PKG=ec2

...
```

### CODE REVIEW

File: internal/service/appconfig/configuration_profile.go [hunk 1]
```diff
@@ -63,6 +63,12 @@ func ResourceConfigurationProfile() *schema.Resource {
 				Optional:     true,
 				ValidateFunc: verify.ValidARN,
 			},
+			"type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "AWS.Freeform",
```
Comment (ryancormack) [inline internal/service/appconfig/configuration_profile.go:69]: This default isn't explicitly documented by AWS, but is the observed behaviour

**Comment by ryancormack** [comment]
I think this change would satisfy https://github.com/hashicorp/terraform-provider-aws/issues/23720

**Comment by ryancormack** [comment]
This is my first contribution to Terraform. I think I've ticked all the contributing requirements. If I've not got something quite right, I'm more than happy to fix a mistake. This is also my first time writing Go. The PR seems small, and I've followed the existing work.

My only thought/consideration was using the default. This is the existing behaviour that happens on the AWS API. I'm unsure whether I should be exposing the default, or leaving it empty, and allowing the AWS API to take care of applying a default.

Thanks

File: internal/service/appconfig/configuration_profile.go (hunk 1 above)
Comment (ewbankkit) [reply internal/service/appconfig/configuration_profile.go:69]: That's fine; there are plenty of other resource attributes where we have "discovered" defaults. Thanks.

**Comment by ewbankkit** [comment]
@ryancormack Thanks for the contribution :tada: :clap:.
In general the changes were great.
In order to get this PR merged for this week's release I went ahead and made a few changes:

* Changed `type` to `ForceNew` as it can't be updated in-place
* Set the `type` value in the resource Read method
* Added a local "enumeration" for the valid type values

**Review by ewbankkit — DISMISSED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccAppConfigConfigurationProfile_ PKG=appconfig ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/appconfig/... -v -count 1 -parallel 3 -run='TestAccAppConfigConfigurationProfile_'  -timeout 180m
=== RUN   TestAccAppConfigConfigurationProfile_basic
=== PAUSE TestAccAppConfigConfigurationProfile_basic
=== RUN   TestAccAppConfigConfigurationProfile_disappears
=== PAUSE TestAccAppConfigConfigurationProfile_disappears
=== RUN   TestAccAppConfigConfigurationProfile_Validators_json
=== PAUSE TestAccAppConfigConfigurationProfile_Validators_json
=== RUN   TestAccAppConfigConfigurationProfile_Validators_lambda
=== PAUSE TestAccAppConfigConfigurationProfile_Validators_lambda
=== RUN   TestAccAppConfigConfigurationProfile_Validators_multiple
=== PAUSE TestAccAppConfigConfigurationProfile_Validators_multiple
=== RUN   TestAccAppConfigConfigurationProfile_updateName
=== PAUSE TestAccAppConfigConfigurationProfile_updateName
=== RUN   TestAccAppConfigConfigurationProfile_updateDescription
=== PAUSE TestAccAppConfigConfigurationProfile_updateDescription
=== RUN   TestAccAppConfigConfigurationProfile_tags
=== PAUSE TestAccAppConfigConfigurationProfile_tags
=== CONT  TestAccAppConfigConfigurationProfile_basic
=== CONT  TestAccAppConfigConfigurationProfile_Validators_multiple
=== CONT  TestAccAppConfigConfigurationProfile_tags
--- PASS: TestAccAppConfigConfigurationProfile_basic (19.45s)
=== CONT  TestAccAppConfigConfigurationProfile_updateDescription
--- PASS: TestAccAppConfigConfigurationProfile_tags (44.72s)
=== CONT  TestAccAppConfigConfigurationProfile_updateName
--- PASS: TestAccAppConfigConfigurationProfile_Validators_multiple (47.92s)
=== CONT  TestAccAppConfigConfigurationProfile_Validators_json
--- PASS: TestAccAppConfigConfigurationProfile_updateDescription (31.96s)
=== CONT  TestAccAppConfigConfigurationProfile_Validators_lambda
--- PASS: TestAccAppConfigConfigurationProfile_updateName (29.24s)
=== CONT  TestAccAppConfigConfigurationProfile_disappears
--- PASS: TestAccAppConfigConfigurationProfile_disappears (15.54s)
--- PASS: TestAccAppConfigConfigurationProfile_Validators_json (42.86s)
--- PASS: TestAccAppConfigConfigurationProfile_Validators_lambda (78.26s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/appconfig	134.004s
```

**Comment by ryancormack** [comment]
Thanks @ewbankkit .

I'm just testing the various behaviours with this App Config option locally now. I'm discovering some 'quirks' with the API. Some of the AWS docs don't appear to be fully complete, so I want to make sure I'm not doing anything unintended.

Should I close the PR for this and re-open as a draft?

**Comment by ewbankkit** [comment]
@ryancormack You can use this PR to make additional changes; I dismissed the Approval.

**Comment by ryancormack** [comment]
@ewbankkit The behaviour I was unsure about was fixed by your change to enable `ForceNew` - learning more new things there.

I've added some documentation to show example schemas for the two types of configuration profile versions now too.

I think it's finished from me and is behaving as intended after that ForceNew.

Thanks very much

**Pull request merged**